sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala

```diff
@@ -247,31 +254,45 @@ class GpuParquetWriter(
    */
   override def write(batch: ColumnarBatch,
      statsTrackers: Seq[ColumnarWriteTaskStatsTracker]): Unit = {
+    val outputMillis = outputTimestampType == ParquetOutputTimestampType.TIMESTAMP_MILLIS.toString
     val newBatch =
-      if (outputTimestampType == ParquetOutputTimestampType.TIMESTAMP_MILLIS.toString) {
-        new ColumnarBatch(GpuColumnVector.extractColumns(batch).map {
-          cv => {
-            cv.dataType() match {
-              case DataTypes.TimestampType => new GpuColumnVector(DataTypes.TimestampType,
-                withResource(cv.getBase()) { v =>
-                  v.castTo(DType.TIMESTAMP_MILLISECONDS)
-                })
-              case _ => cv
-            }
+      new ColumnarBatch(GpuColumnVector.extractColumns(batch).map {
+        cv => {
+          cv.dataType() match {
+            case DataTypes.TimestampType if outputMillis =>
+              new GpuColumnVector(DataTypes.TimestampType, withResource(cv.getBase()) { v =>
+                v.castTo(DType.TIMESTAMP_MILLISECONDS)
+              })
+            case d: DecimalType if d.precision < 10 =>
+              // There is a bug in Spark that causes a problem if we write Decimals with
+              // precision < 10 as Decimal64.
+              // https://issues.apache.org/jira/browse/SPARK-34167
+              new GpuColumnVector(d, withResource(cv.getBase()) { v =>
+                v.castTo(DType.create(DType.DTypeEnum.DECIMAL32, -d.scale))
+              })
+            case _ => cv
           }
-        })
-      } else {
-        batch
-      }
+        }
+      })
 
     super.write(newBatch, statsTrackers)
   }
   override val tableWriter: TableWriter = {
+    def precisionsList(t: DataType): List[Int] = {
+      t match {
+        case d: DecimalType => List(d.precision)
+        case s: StructType => s.flatMap(f => precisionsList(f.dataType)).toList
```

Comment:
we could save the conversion `toList` if the return type for `precisionsList` were `Seq[Int]`